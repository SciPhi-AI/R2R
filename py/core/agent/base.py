import asyncio
from abc import ABCMeta
from typing import AsyncGenerator, Generator, Optional

from core.base.abstractions import (
    AsyncSyncMeta,
    LLMChatCompletion,
    LLMChatCompletionChunk,
    Message,
    syncable,
)
from core.base.agent import Agent, Conversation
import logging

logger = logging.getLogger(__name__)


class CombinedMeta(AsyncSyncMeta, ABCMeta):
    pass


def sync_wrapper(async_gen):
    loop = asyncio.get_event_loop()

    def wrapper():
        try:
            while True:
                try:
                    yield loop.run_until_complete(async_gen.__anext__())
                except StopAsyncIteration:
                    break
        finally:
            loop.run_until_complete(async_gen.aclose())

    return wrapper()


class R2RAgent(Agent, metaclass=CombinedMeta):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._register_tools()
        self._reset()

    def _reset(self):
        self._completed = False
        self.conversation = Conversation()

    @syncable
    async def arun(
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
        *args,
        **kwargs,
    ) -> list[dict]:
        self._reset()
        await self._setup(system_instruction)

        if messages:
            for message in messages:
                await self.conversation.add_message(message)

        while not self._completed:
            messages_list = await self.conversation.get_messages()
            generation_config = self.get_generation_config(messages_list[-1])
            response = await self.llm_provider.aget_completion(
                messages_list,
                generation_config,
            )
            await self.process_llm_response(response, *args, **kwargs)

        return await self.conversation.get_messages()

    async def process_llm_response(
        self, response: LLMChatCompletion, *args, **kwargs
    ) -> None:
        if not self._completed:
            message = response.choices[0].message
            if message.function_call:
                await self.handle_function_or_tool_call(
                    message.function_call.name,
                    message.function_call.arguments,
                    *args,
                    **kwargs,
                )
            elif message.tool_calls:
                for tool_call in message.tool_calls:
                    await self.handle_function_or_tool_call(
                        tool_call.function.name,
                        tool_call.function.arguments,
                        *args,
                        **kwargs,
                    )
            else:
                await self.conversation.add_message(
                    Message(role="assistant", content=message.content)
                )
                self._completed = True


class R2RStreamingAgent(R2RAgent):
    async def arun(  # type: ignore
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
        *args,
        **kwargs,
    ) -> AsyncGenerator[str, None]:
        self._reset()
        await self._setup(system_instruction)

        if messages:
            for message in messages:
                await self.conversation.add_message(message)

        while not self._completed:
            messages_list = await self.conversation.get_messages()

            generation_config = self.get_generation_config(
                messages_list[-1], stream=True
            )
            stream = self.llm_provider.get_completion_stream(
                messages_list,
                generation_config,
            )
            async for chunk in self.process_llm_response(
                stream, *args, **kwargs
            ):
                yield chunk

    def run(
        self, system_instruction, messages, *args, **kwargs
    ) -> Generator[str, None, None]:
        return sync_wrapper(
            self.arun(system_instruction, messages, *args, **kwargs)
        )

    async def process_llm_response(  # type: ignore
        self,
        stream: Generator[LLMChatCompletionChunk, None, None],
        *args,
        **kwargs,
    ) -> AsyncGenerator[str, None]:
        function_name = None
        function_arguments = ""
        content_buffer = ""

        for chunk in stream:
            delta = chunk.choices[0].delta
            if delta.tool_calls:
                for tool_call in delta.tool_calls:
                    if not tool_call.function:
                        logger.info("Tool function not found in tool call.")
                        continue
                    name = tool_call.function.name
                    if not name:
                        logger.info("Tool name not found in tool call.")
                        continue
                    arguments = tool_call.function.arguments
                    if not arguments:
                        logger.info("Tool arguments not found in tool call.")
                        continue

                    results = await self.handle_function_or_tool_call(
                        name,
                        arguments,
                        # FIXME: tool_call.id,
                        *args,
                        **kwargs,
                    )

                    yield "<tool_call>"
                    yield f"<name>{name}</name>"
                    yield f"<arguments>{arguments}</arguments>"
                    yield f"<results>{results.llm_formatted_result}</results>"
                    yield "</tool_call>"

            if delta.function_call:
                if delta.function_call.name:
                    function_name = delta.function_call.name
                if delta.function_call.arguments:
                    function_arguments += delta.function_call.arguments
            elif delta.content:
                if content_buffer == "":
                    yield "<completion>"
                content_buffer += delta.content
                yield delta.content

            if chunk.choices[0].finish_reason == "function_call":
                if not function_name:
                    logger.info("Function name not found in function call.")
                    continue

                yield "<function_call>"
                yield f"<name>{function_name}</name>"
                yield f"<arguments>{function_arguments}</arguments>"
                tool_result = await self.handle_function_or_tool_call(
                    function_name, function_arguments, *args, **kwargs
                )
                if tool_result.stream_result:
                    yield f"<results>{tool_result.stream_result}</results>"
                else:
                    yield f"<results>{tool_result.llm_formatted_result}</results>"

                yield "</function_call>"

                function_name = None
                function_arguments = ""

            elif chunk.choices[0].finish_reason == "stop":
                if content_buffer:
                    await self.conversation.add_message(
                        Message(role="assistant", content=content_buffer)
                    )
                self._completed = True
                yield "</completion>"

        # Handle any remaining content after the stream ends
        if content_buffer and not self._completed:
            await self.conversation.add_message(
                Message(role="assistant", content=content_buffer)
            )
            self._completed = True
            yield "</completion>"
