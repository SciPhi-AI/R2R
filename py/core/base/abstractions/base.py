import asyncio


class AsyncSyncMeta(type):
    _event_loop = None  # Class-level shared event loop

    @classmethod
    def get_event_loop(cls):
        if cls._event_loop is None or cls._event_loop.is_closed():
            cls._event_loop = asyncio.new_event_loop()
            asyncio.set_event_loop(cls._event_loop)
        return cls._event_loop

    def __new__(cls, name, bases, dct):
        new_cls = super().__new__(cls, name, bases, dct)
        for attr_name, attr_value in dct.items():
            if asyncio.iscoroutinefunction(attr_value) and getattr(
                attr_value, "_syncable", False
            ):
                sync_method_name = attr_name[
                    1:
                ]  # Remove leading 'a' for sync method
                async_method = attr_value

                def make_sync_method(async_method):
                    def sync_wrapper(self, *args, **kwargs):
                        loop = cls.get_event_loop()
                        if not loop.is_running():
                            # Setup to run the loop in a background thread if necessary
                            # to prevent blocking the main thread in a synchronous call environment
                            from threading import Thread

                            result = None
                            exception = None

                            def run():
                                nonlocal result, exception
                                try:
                                    asyncio.set_event_loop(loop)
                                    result = loop.run_until_complete(
                                        async_method(self, *args, **kwargs)
                                    )
                                except Exception as e:
                                    exception = e
                                finally:
                                    generation_config = kwargs.get(
                                        "rag_generation_config", None
                                    )
                                    if (
                                        not generation_config
                                        or not generation_config.stream
                                    ):
                                        loop.run_until_complete(
                                            loop.shutdown_asyncgens()
                                        )
                                        loop.close()

                            thread = Thread(target=run)
                            thread.start()
                            thread.join()
                            if exception:
                                raise exception
                            return result
                        else:
                            # If there's already a running loop, schedule and execute the coroutine
                            future = asyncio.run_coroutine_threadsafe(
                                async_method(self, *args, **kwargs), loop
                            )
                            return future.result()

                    return sync_wrapper

                setattr(
                    new_cls, sync_method_name, make_sync_method(async_method)
                )
        return new_cls


def syncable(func):
    """Decorator to mark methods for synchronous wrapper creation."""
    func._syncable = True
    return func
