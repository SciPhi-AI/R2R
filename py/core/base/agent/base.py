from typing import Any, Callable, Dict, Optional

from ..abstractions import R2RSerializable


class Tool(R2RSerializable):
    name: str
    description: str
    results_function: Callable
    llm_format_function: Callable
    stream_function: Optional[Callable] = None
    parameters: Optional[Dict[str, Any]] = None

    class Config:
        arbitrary_types_allowed = True


class ToolResult(R2RSerializable):
    raw_result: Any
    llm_formatted_result: str
    stream_result: Optional[str] = None
