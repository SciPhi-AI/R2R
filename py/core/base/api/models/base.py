from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class ResultsWrapper(BaseModel, Generic[T]):
    results: T
