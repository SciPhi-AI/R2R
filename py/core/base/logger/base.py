import logging
from abc import abstractmethod
from datetime import datetime
from enum import Enum
from typing import Any, Optional, Tuple, Union
from uuid import UUID

from pydantic import BaseModel

from core.base import Message

from ..providers.base import Provider, ProviderConfig

logger = logging.getLogger()


class RunInfoLog(BaseModel):
    run_id: UUID
    run_type: str
    timestamp: datetime
    user_id: UUID


class PersistentLoggingConfig(ProviderConfig):
    provider: str = "local"
    log_table: str = "logs"
    log_info_table: str = "log_info"
    logging_path: Optional[str] = None

    def validate_config(self) -> None:
        pass

    @property
    def supported_providers(self) -> list[str]:
        return ["local", "postgres"]


class RunType(str, Enum):
    """Enumeration of the different types of runs."""

    RETRIEVAL = "RETRIEVAL"
    MANAGEMENT = "MANAGEMENT"
    INGESTION = "INGESTION"
    AUTH = "AUTH"
    UNSPECIFIED = "UNSPECIFIED"
    KG = "KG"


class PersistentLoggingProvider(Provider):
    @abstractmethod
    async def close(self):
        pass

    @abstractmethod
    async def log(
        self,
        run_id: UUID,
        key: str,
        value: str,
    ):
        pass

    @abstractmethod
    async def get_logs(
        self,
        run_ids: list[UUID],
        limit_per_run: int,
    ) -> list:
        pass

    @abstractmethod
    async def info_log(
        self,
        run_id: UUID,
        run_type: RunType,
        user_id: UUID,
    ):
        pass

    @abstractmethod
    async def get_info_logs(
        self,
        offset: int = 0,
        limit: int = 100,
        run_type_filter: Optional[RunType] = None,
        user_ids: Optional[list[UUID]] = None,
    ) -> list[RunInfoLog]:
        pass
