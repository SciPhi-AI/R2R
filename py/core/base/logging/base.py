from enum import Enum


class RunType(str, Enum):
    """Enumeration of the different types of runs."""

    RETRIEVAL = "RETRIEVAL"
    MANAGEMENT = "MANAGEMENT"
    INGESTION = "INGESTION"
    AUTH = "AUTH"
    UNSPECIFIED = "UNSPECIFIED"
    KG = "KG"
