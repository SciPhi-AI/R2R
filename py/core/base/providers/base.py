from abc import ABC, abstractmethod
from typing import Any, Optional, Type

from pydantic import BaseModel

from ..abstractions import R2RSerializable


class AppConfig(R2RSerializable):
    project_name: Optional[str] = None

    @classmethod
    def create(cls, *args, **kwargs):
        project_name = kwargs.get("project_name")
        return AppConfig(project_name=project_name)


class ProviderConfig(BaseModel, ABC):
    """A base provider configuration class"""

    app: AppConfig  # Add an app_config field
    extra_fields: dict[str, Any] = {}
    provider: Optional[str] = None

    class Config:
        arbitrary_types_allowed = True
        ignore_extra = True

    @abstractmethod
    def validate_config(self) -> None:
        pass

    @classmethod
    def create(cls: Type["ProviderConfig"], **kwargs: Any) -> "ProviderConfig":
        base_args = cls.model_fields.keys()
        filtered_kwargs = {
            k: v if v != "None" else None
            for k, v in kwargs.items()
            if k in base_args
        }
        instance = cls(**filtered_kwargs)
        for k, v in kwargs.items():
            if k not in base_args:
                instance.extra_fields[k] = v
        return instance

    @property
    @abstractmethod
    def supported_providers(self) -> list[str]:
        """Define a list of supported providers."""
        pass

    @classmethod
    def from_dict(
        cls: Type["ProviderConfig"], data: dict[str, Any]
    ) -> "ProviderConfig":
        """Create a new instance of the config from a dictionary."""
        return cls.create(**data)


class Provider(ABC):
    """A base provider class to provide a common interface for all providers."""

    def __init__(self, config: ProviderConfig, *args, **kwargs):
        if config:
            config.validate_config()
        self.config = config
