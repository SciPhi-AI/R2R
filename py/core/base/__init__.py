from .abstractions import *
from .agent import *
from .api.models import *
from .logging import *
from .parsers import *
from .pipeline import *
from .pipes import *
from .providers import *
from .utils import *

__all__ = [
    ## ABSTRACTIONS
    # Base abstractions
    "AsyncSyncMeta",
    "syncable",
    # Completion abstractions
    "CompletionRecord",
    "MessageType",
    # Document abstractions
    "DataType",
    "Document",
    "DocumentExtraction",
    "DocumentFragment",
    "DocumentInfo",
    "IngestionStatus",
    "RestructureStatus",
    "DocumentType",
    # Embedding abstractions
    "EmbeddingPurpose",
    "default_embedding_prefixes",
    # Exception abstractions
    "R2RDocumentProcessingError",
    "R2RException",
    # KG abstractions
    "Entity",
    "KGExtraction",
    "Triple",
    # LLM abstractions
    "GenerationConfig",
    "LLMChatCompletion",
    "LLMChatCompletionChunk",
    "RAGCompletion",
    # Prompt abstractions
    "Prompt",
    # Search abstractions
    "AggregateSearchResult",
    "KGLocalSearchResult",
    "KGGlobalSearchResult",
    "KGSearchResult",
    "KGSearchSettings",
    "VectorSearchResult",
    "VectorSearchSettings",
    # Restructure abstractions
    "KGCreationSettings",
    "KGEnrichmentSettings",
    # User abstractions
    "Token",
    "TokenData",
    "UserStats",
    # Vector abstractions
    "Vector",
    "VectorEntry",
    "VectorType",
    "StorageResult",
    ## AGENT
    # Agent abstractions
    "Agent",
    "AgentConfig",
    "Conversation",
    "Message",
    "Tool",
    "ToolResult",
    ## API
    # Auth Responses
    "GenericMessageResponse",
    "TokenResponse",
    "UserResponse",
    ## LOGGING
    # Basic types
    "RunType",
    "AnalysisTypes",
    "LogAnalytics",
    "LogAnalyticsConfig",
    "LogFilterCriteria",
    "LogProcessor",
    # Logging Providers
    "LocalRunLoggingProvider",
    "LoggingConfig",
    "PostgresLoggingConfig",
    "PostgresRunLoggingProvider",
    "RedisLoggingConfig",
    "RedisRunLoggingProvider",
    "RunLoggingSingleton",
    # Run Manager
    "RunManager",
    "manage_run",
    ## PARSERS
    # Base parser
    "AsyncParser",
    ## PIPELINE
    # Base pipeline
    "AsyncPipeline",
    ## PIPES
    "AsyncPipe",
    "AsyncState",
    "PipeType",
    ## PROVIDERS
    # Base provider classes
    "Provider",
    "ProviderConfig",
    # Auth provider
    "AuthConfig",
    "AuthProvider",
    # Chunking provider
    "R2RChunkingConfig",
    "UnstructuredChunkingConfig",
    "ChunkingConfig",
    "ChunkingProvider",
    "Strategy",
    # Crypto provider
    "CryptoConfig",
    "CryptoProvider",
    # Database providers
    "DatabaseConfig",
    "DatabaseProvider",
    "RelationalDBProvider",
    "VectorDBProvider",
    # Embedding provider
    "EmbeddingConfig",
    "EmbeddingProvider",
    # File provider
    "FileConfig",
    "FileProvider",
    # Knowledge Graph provider
    "KGConfig",
    "KGProvider",
    # LLM provider
    "CompletionConfig",
    "CompletionProvider",
    # Parsing provider
    "ParsingConfig",
    "ParsingProvider",
    "OverrideParser",
    # Prompt provider
    "PromptConfig",
    "PromptProvider",
    ## UTILS
    "RecursiveCharacterTextSplitter",
    "TextSplitter",
    "run_pipeline",
    "to_async_generator",
    "generate_run_id",
    "generate_id_from_label",
    "increment_version",
    "EntityType",
    "RelationshipType",
    "format_entity_types",
    "format_relations",
]
