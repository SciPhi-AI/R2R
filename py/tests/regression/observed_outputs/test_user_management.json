{
  "register_user": {
    "results": {
      "id": "45c36a22-6c43-56a7-b319-9abaae3f1bc0",
      "email": "test_6b09cf57@example.com",
      "is_active": true,
      "is_superuser": false,
      "created_at": "2024-10-03T23:00:58.216612Z",
      "updated_at": "2024-10-03T23:00:58.216612Z",
      "is_verified": false,
      "collection_ids": [],
      "hashed_password": "$2b$12$TSiUhA7qXile9sAIwsde0uDFXuAitx5l7d79zq7H8pQitGgl1lmMi",
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null
    }
  },
  "login_user": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzZiMDljZjU3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI4MDAwMDU4LjUxMjY4NCwidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.EnGuTyqILBZFTZdI22mYXjuNs0KsC03Mu7wDh8I6VUE",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzZiMDljZjU3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI4NjAxMjU4LCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.twvIBFxz_I8nODyNcWW8keJvHL7PEFBieqLm18pxol8",
        "token_type": "refresh"
      }
    }
  },
  "user_info": {
    "results": {
      "id": "45c36a22-6c43-56a7-b319-9abaae3f1bc0",
      "email": "test_6b09cf57@example.com",
      "is_active": true,
      "is_superuser": false,
      "created_at": "2024-10-03T23:00:58.216612Z",
      "updated_at": "2024-10-03T23:00:58.216612Z",
      "is_verified": true,
      "collection_ids": [
        "20efadf0-b06f-5685-a969-ba2800d8fee4"
      ],
      "hashed_password": "$2b$12$TSiUhA7qXile9sAIwsde0uDFXuAitx5l7d79zq7H8pQitGgl1lmMi",
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null
    }
  },
  "change_password": {
    "results": {
      "message": "Password changed successfully"
    }
  },
  "update_profile": {
    "results": "AuthMethods.update_user() missing 1 required positional argument: 'user_id'"
  },
  "refresh_token": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzZiMDljZjU3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI4MDAwMDU5LjAzNjY0MywidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.jqD5fmtx_FP6OGA8N2cjIWx9lD-MBVqP1EnY564ecVM",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzZiMDljZjU3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI4NjAxMjU5LCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.dQKCMP8-irRaZ0NiP7Pkh6aQdHDwG_ZTyIDiLFZV9vw",
        "token_type": "refresh"
      }
    }
  },
  "superuser_test": {
    "results": [
      {
        "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
        "email": "admin@example.com",
        "is_superuser": true,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-10-03T16:58:58.104181Z",
        "updated_at": "2024-10-03T16:58:58.104181Z",
        "collection_ids": [
          "122fdf6a-e116-546b-a8f6-e4cb2e2c0a09"
        ],
        "num_files": 15,
        "total_size_in_bytes": 8862496,
        "document_ids": [
          "5b1bd54f-4d70-56b9-a017-a618bc75f94c",
          "30f950f0-c692-57c5-b6ec-ff78ccf5ccdc",
          "2f576170-c4f9-5141-a910-a0924f341de4",
          "3e157b3a-8469-51db-90d9-52e7d896b49b",
          "451adbbd-e24e-5c9b-80a8-f424b6c7accb",
          "d421207a-d799-5806-8d67-46b2005b15d4",
          "db02076e-989a-59cd-98d5-e24e15a0bd27",
          "716fea3a-826b-5b27-8e59-ffbd1a35455a",
          "7b0f40c5-2ace-5781-ae35-ead99ddee8c5",
          "c3291abf-8a4e-5d9d-80fd-232ef6fd8526",
          "e797da22-8c5d-54e5-bed5-a55954cf6bf9",
          "7c1105fc-8f62-5a8f-ac81-fe88f3ec9e4c",
          "57eec3df-cf68-5559-a80d-ae3fb55b9af1",
          "01d514a4-5cb1-5c86-be8c-0c5c312a02c9",
          "52e12576-090f-59db-91f4-6d4b2e29ae6c"
        ]
      },
      {
        "user_id": "45c36a22-6c43-56a7-b319-9abaae3f1bc0",
        "email": "test_6b09cf57@example.com",
        "is_superuser": false,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-10-03T23:00:58.216612Z",
        "updated_at": "2024-10-03T23:00:59.019061Z",
        "collection_ids": [
          "20efadf0-b06f-5685-a969-ba2800d8fee4"
        ],
        "num_files": 0,
        "total_size_in_bytes": 0,
        "document_ids": []
      }
    ],
    "total_entries": 2
  },
  "logout": {
    "results": {
      "message": "Logged out successfully"
    }
  },
  "delete_account": {
    "results": {
      "message": "User account f7495bfb-58e1-539e-8b8d-6cb61f821c2b deleted successfully."
    }
  }
}
