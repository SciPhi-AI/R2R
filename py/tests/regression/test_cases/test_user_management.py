import uuid

from tests.regression.test_cases.base import BaseTest


class TestUserManagement(BaseTest):
    VARIABLE_USER_FIELDS = [
        "id",
        "email",
        "created_at",
        "updated_at",
        "hashed_password",
        "access_token",
        "refresh_token",
    ]

    def __init__(self, client):
        super().__init__(client)
        self.set_exclude_paths(
            "register_user",
            [
                f"root['{field}']"
                for field in TestUserManagement.VARIABLE_USER_FIELDS
            ],
        )
        self.set_exclude_paths(
            "login_user",
            [
                f"root['{field}']"
                for field in ["access_token", "refresh_token"]
            ],
        )
        self.set_exclude_paths(
            "user_info",
            [
                f"root['{field}']"
                for field in TestUserManagement.VARIABLE_USER_FIELDS
            ],
        )
        self.set_exclude_paths(
            "change_password",
            [
                f"root['{field}']"
                for field in TestUserManagement.VARIABLE_USER_FIELDS
            ],
        )
        self.set_exclude_paths(
            "update_profile",
            [
                f"root['{field}']"
                for field in TestUserManagement.VARIABLE_USER_FIELDS
            ],
        )
        self.set_exclude_paths(
            "refresh_token",
            [
                f"root['{field}']"
                for field in TestUserManagement.VARIABLE_USER_FIELDS
            ],
        )
        self.user_id_string = str(uuid.uuid4()).split("-")[0]
        self.user_id_string_2 = str(uuid.uuid4()).split("-")[0]

    def get_test_cases(self):
        return {
            "register_user": lambda client: self.register_user_test(client),
            "login_user": lambda client: self.login_user_test(client),
            "user_info": lambda client: self.user_info_test(client),
            "change_password": lambda client: self.change_password_test(
                client
            ),
            # "reset_password": lambda client: self.reset_password_test(client),
            "update_profile": lambda client: self.update_profile_test(client),
            "refresh_token": lambda client: self.refresh_token_test(client),
            "superuser_test": lambda client: self.superuser_test(client),
            "logout": lambda client: self.logout_test(client),
            "delete_account": lambda client: self.delete_user_test(client),
            "login_user": lambda client: self.login_user_test(client),
            "refresh_token": lambda client: self.refresh_token_test(client),
        }

    def register_user_test(self, client):
        try:
            email = f"test_{self.user_id_string}@example.com"
            password = "password123"
            user = client.register(email, password)
            self.user = user
            return user
        except Exception as e:
            return {"results": str(e)}

    def login_user_test(self, client):
        try:
            email = f"test_{self.user_id_string}@example.com"
            password = "password123"
            login = client.login(email, password)
            return login
        except Exception as e:
            return {"results": str(e)}

    def user_info_test(self, client):
        try:
            return client.user()
        except Exception as e:
            return {"results": str(e)}

    def change_password_test(self, client):
        try:
            return client.change_password("password123", "new_password")
        except Exception as e:
            return {"results": str(e)}

    # def reset_password_test(self, client):
    #     try:
    #         reset_request = client.request_password_reset("test@example.com")
    #         # In a real scenario, we'd need to get the reset token from the email
    #         reset_token = "mock_reset_token"
    #         return client.confirm_password_reset(reset_token, "new_password")
    #     except Exception as e:
    #         return {"results": str(e)}

    def update_profile_test(self, client):
        try:
            return client.update_user(name="John Doe", bio="R2R enthusiast")
        except Exception as e:
            return {"results": str(e)}

    def delete_user_test(self, client):
        try:
            email = f"test_{self.user_id_string_2}@example.com"
            password = "password123"
            user = client.register(email, password)

            return client.delete_user(user["results"]["id"], "password123")
        except Exception as e:
            return {"results": str(e)}

    def refresh_token_test(self, client):
        try:
            return client.refresh_access_token()
        except Exception as e:
            return {"results": str(e)}

    def superuser_test(self, client):
        try:
            # Login as admin
            client.login("admin@example.com", "change_me_immediately")
            return client.users_overview()
        except Exception as e:
            return {"results": str(e)}

    def logout_test(self, client):
        try:
            return client.logout()
        except Exception as e:
            return {"results": str(e)}
