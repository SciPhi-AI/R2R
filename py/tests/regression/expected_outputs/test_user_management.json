{
  "register_user": {
    "results": {
      "id": "14466e62-fd0a-5ad3-85b6-6965d7f6d336",
      "email": "test_4f3ba5b7@example.com",
      "is_active": true,
      "is_superuser": false,
      "created_at": "2024-09-20T22:48:31.575356Z",
      "updated_at": "2024-09-20T22:48:31.575356Z",
      "is_verified": false,
      "collection_ids": [],
      "hashed_password": "$2b$12$91hUOrNh1OhZZODomjxa0Oc/hGNUVwxFh45CXXBBOfhQP98y.pvFm",
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null
    }
  },
  "login_user": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzRmM2JhNWI3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI2ODc2MTExLjgzOTksInRva2VuX3R5cGUiOiJhY2Nlc3MifQ.vW27WohcKi9ipcTC04i6rSg42ZH8NmnXFYUgfXcGAw0",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzRmM2JhNWI3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI3NDc3MzExLCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.NUBrdq0PFM8TbDN4EGEefvke07HMLp0T4W6zUbOxt7w",
        "token_type": "refresh"
      }
    }
  },
  "user_info": {
    "results": {
      "id": "14466e62-fd0a-5ad3-85b6-6965d7f6d336",
      "email": "test_4f3ba5b7@example.com",
      "is_active": true,
      "is_superuser": false,
      "created_at": "2024-09-20T22:48:31.575356Z",
      "updated_at": "2024-09-20T22:48:31.575356Z",
      "is_verified": true,
      "collection_ids": [],
      "hashed_password": "$2b$12$91hUOrNh1OhZZODomjxa0Oc/hGNUVwxFh45CXXBBOfhQP98y.pvFm",
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null
    }
  },
  "change_password": {
    "results": {
      "message": "Password changed successfully"
    }
  },
  "update_profile": {
    "results": {
      "id": "14466e62-fd0a-5ad3-85b6-6965d7f6d336",
      "email": "test_4f3ba5b7@example.com",
      "is_active": true,
      "is_superuser": false,
      "created_at": "2024-09-20T22:48:31.575356Z",
      "updated_at": "2024-09-20T22:48:32.394281Z",
      "is_verified": true,
      "collection_ids": [],
      "hashed_password": null,
      "verification_code_expiry": null,
      "name": "John Doe",
      "bio": "R2R enthusiast",
      "profile_picture": null
    }
  },
  "refresh_token": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzRmM2JhNWI3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI2ODc2MTEyLjQyMjYzNywidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.09C6tAMd3WhrEmsqPdFsWx7lxsp0abGSdWGessddSw0",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0XzRmM2JhNWI3QGV4YW1wbGUuY29tIiwiZXhwIjoxNzI3NDc3MzEyLCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.yrm_ncSzAScPpdlqYHD2WWhy1AcwuO9CGHcrNLrTNGY",
        "token_type": "refresh"
      }
    }
  },
  "superuser_test": {
    "results": [
      {
        "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
        "email": "admin@example.com",
        "is_superuser": true,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-09-20T22:26:10.261930Z",
        "updated_at": "2024-09-20T22:26:10.261930Z",
        "collection_ids": [],
        "num_files": 15,
        "total_size_in_bytes": 8767226,
        "document_ids": [
          "57eec3df-cf68-5559-a80d-ae3fb55b9af1",
          "7c1105fc-8f62-5a8f-ac81-fe88f3ec9e4c",
          "451adbbd-e24e-5c9b-80a8-f424b6c7accb",
          "01d514a4-5cb1-5c86-be8c-0c5c312a02c9",
          "e797da22-8c5d-54e5-bed5-a55954cf6bf9",
          "5b1bd54f-4d70-56b9-a017-a618bc75f94c",
          "30f950f0-c692-57c5-b6ec-ff78ccf5ccdc",
          "52e12576-090f-59db-91f4-6d4b2e29ae6c",
          "7b0f40c5-2ace-5781-ae35-ead99ddee8c5",
          "c3291abf-8a4e-5d9d-80fd-232ef6fd8526",
          "d421207a-d799-5806-8d67-46b2005b15d4",
          "9fbe403b-c11c-5aae-8ade-ef22980c3ad1",
          "716fea3a-826b-5b27-8e59-ffbd1a35455a",
          "2f576170-c4f9-5141-a910-a0924f341de4",
          "3e157b3a-8469-51db-90d9-52e7d896b49b"
        ]
      },
      {
        "user_id": "a0a60c19-7672-5537-a71c-07a771bc4fa3",
        "email": "test_40669f1c@example.com",
        "is_superuser": false,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-09-20T22:48:01.453932Z",
        "updated_at": "2024-09-20T22:48:01.453932Z",
        "collection_ids": [],
        "num_files": 0,
        "total_size_in_bytes": 0,
        "document_ids": []
      },
      {
        "user_id": "14466e62-fd0a-5ad3-85b6-6965d7f6d336",
        "email": "test_4f3ba5b7@example.com",
        "is_superuser": false,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-09-20T22:48:31.575356Z",
        "updated_at": "2024-09-20T22:48:32.394281Z",
        "collection_ids": [],
        "num_files": 0,
        "total_size_in_bytes": 0,
        "document_ids": []
      },
      {
        "user_id": "a5e4ef83-b9ae-56e0-952b-4116f487b000",
        "email": "test_c8a21611@example.com",
        "is_superuser": false,
        "is_active": true,
        "is_verified": true,
        "created_at": "2024-09-20T22:48:14.196002Z",
        "updated_at": "2024-09-20T22:48:14.196002Z",
        "collection_ids": [],
        "num_files": 0,
        "total_size_in_bytes": 0,
        "document_ids": []
      }
    ],
    "total_entries": 4
  },
  "logout": {
    "results": {
      "message": "Logged out successfully"
    }
  },
  "delete_account": {
    "results": {
      "message": "User account c8c03ba0-6d1d-524c-be39-7fb6e01b4298 deleted successfully."
    }
  }
}
